### ISSUE TITLE
Can we embed the whiteboard into a ReactJS App

### ISSUE THREAD

**Issue by yashrajpahwa** [tableaunoir/tableaunoir#189]
Hi,

Is there a JS library to add the whiteboard to a react app. I need only the front end for my app.
Is there an API that has events such as onChange & onUpdate which will allow me to integrate it with my own API

Thanks

**Comment by francoisschwarzentruber** [comment]
Dear @yashrajpahwa, thank you for having posted your interesting issue! not yet! I do not react app very well. What are the desiderata? Like, a tableaunoirlibrary.js to include, and one adds a div - let say named "board", and then we do sth like this?

   `let T = tableaunoirlibrary.make(board)`
   `T.onUpdate( () => ...)`

**Comment by yashrajpahwa** [comment]
Absolutely, something like the above lines of code. 

- `let T = tableaunoirlibrary.make(board)` - This will manipulate the DOM to add the div element with ID board.
- `T.addControls({id: "controls", menu: false, pen: true})` - This will allow adding controls to the toolbar. `id` is the element id, menu and pen are tools and this would be the case for all other tools such as the eraser tool
- `T.onUpdate( () => ...)` - triggers when the state of the board is updated
- `T.onErase( () => ...)`- triggers when the board has been erased and so on

Please let me know if I could help in any way @francoisschwarzentruber .

Thanks

**Comment by francoisschwarzentruber** [comment]
I think it is an excellent idea! The main issue I see is that all the layout of tableaunoir is done in index.html and not in js. I think the best way to do would be to transform the architecture of Tableaunoir, so that even tableaunoir.github.io itself just use tableaunoirlibrary :)

**Comment by yashrajpahwa** [comment]
Yes, I can see that you have most of the logic implemented in JS but you are using electron to build the compiled website, you can extract the logic into plain typescript and then make the internal events accessible via an API. You can use a bundler to compile and minify the code and then publish the package on NPM and make the file accessible via a CDN like JSDeliver.

This way the logic will remain in the library and the UI can be handled in the frontend.
If you are ok with the idea you may proceed.

Please add a comment to the conversation when the library is ready and then you can close this issue.

**Comment by francoisschwarzentruber** [comment]
Most of the logic is implemented in Typescript (TS°. Then, the whole project with a process with Webpack and tsc (the Typescript compiler), to produce tableaunoir.js etc.

I am not an expert in making packages. But I think it is no costly to prepare such a tableaunoirlibrary.js, that will be actually the current tableaunoir.js: I will just add a TS file that describes the API and make it accessible via Vanilla JS.

Then I will probably need some help for:
- publishing a package on NPM (I have no idea how to do that yet)
- making accessible via a CDN
- understanding what to do with the assets (Probably they can be put in an NPM package, but I am not sure)

PS: electron is used to launch the desktop version of the application.
PS2: currently, I do not have so many times. I may be slow in solving that issue.

**Comment by francoisschwarzentruber** [comment]
Technical question: do you know how it is possible to elegantly use the html description of a tableaunoir board from JS? For instance, let T = tableaunoirlibrary.make(board) will attach to the DOM element board all the html template for a Tableaunoir board.

**Comment by yashrajpahwa** [comment]
Hi François,

Which file in the source folder is responsible for interacting with the backend WebSocket server. Also, I am not able to find tableaunoir.js which has been imported in src/index.html. 

Need some help with this.

Thanks

**Comment by francoisschwarzentruber** [comment]
Hello,

thank you for your message.
- The source file that interacts with the backend is `share.ts`.
- The file `tableaunoir.js` is built when your run `npm run build`.

**Comment by francoisschwarzentruber** [comment]
Hello,

maybe another solution would be
`<iframe src="https://tableaunoir.github.io" width="1000px" height="500px"></iframe>`
            
What do you think?

**Comment by yashrajpahwa** [comment]
> Hello,
> 
> maybe another solution would be `<iframe src="https://tableaunoir.github.io" width="1000px" height="500px"></iframe>`
> 
> What do you think?

Yes, that will work for the people who want to have a random board to scribble on, but accessing the core API to save the state maybe or restore it programmatically (including other functions) won't be possible with this solution.
